src/freertos_drivers/esp_idf/EspIdfWiFi.cxx

```diff
@@ -340,565 +341,184 @@ void EspIdfWiFiBase::mdns_service_remove(const char *service)
-        std::string name;
-        std::string proto;
-        mdns_split(service, &name, &proto);
-
-        // Enable mDNS on STA if required.
-        bool mdns_adv_inhibited_on_sta = false;
-        if (mdnsAdvInhibitSta_)
+        // The pseudo randomness of the resulting expires time helps
+        // ensure timing jitter between two or more devices using the
+        // same algorithm.
+        long long now = OSTime::get_monotonic();
+        long long expires = MSEC_TO_NSEC(mdnsLookupUd_(mdnsLookupRd_));
+        expires += mdnsLookupTimestamp_;
+        if (expires > now)
         {
-            mdns_adv_inhibited_on_sta = true;
-            mdns_adv_inhibit();
-            mdns_restore_sta();
+            // Ensure some blanking time since the last inhibit.
+            usleep(NSEC_TO_USEC(expires - now));
         }
+        // The placement of this lock must come after the usleep() call above.
+        // this is so that other Wifi API can run concurrently during the sleep.
+        OSMutexLock locker(&lock_);
+        mdnsAdvInhibited_ = true;
+        // Advertising on STA is inhibited, remove all advertising services.
+        ::mdns_service_remove_all();
+        // Enable mDNS on STA.
+        ESP_ERROR_CHECK(
+            mdns_netif_action(staIface_, MDNS_EVENT_ENABLE_IP4));
+    }
 
-        // Do the blocking query for 2 seconds, only 1 result.
-        mdns_result_t *results;
-        esp_err_t err =
-            mdns_query_ptr(name.c_str(), proto.c_str(), 2000, 1, &results);
-
-        // Disable mDNS on STA if required.
-        if (mdns_adv_inhibited_on_sta)
-        {
-            mdns_disable_sta();
-            mdns_adv_inhibit_remove();
-        }
+    std::string name;
+    std::string proto;
+    mdns_split(service, &name, &proto);
 
-        // Check for an error result from the query.
-        if (err != ESP_OK)
-        {
-            LOG_ERROR("wifi: mdns_query_ptr() failed: %s",
-                esp_err_to_name(err));
-                return EAI_AGAIN;
-        }
+    /// @todo In the future, may want to do a faster query which collects more
+    ///       than one result and then tries to asses which of multiple results
+    ///       will be the highest performance option.
+    mdns_result_t *results;
+    esp_err_t err =
+        mdns_query_ptr(name.c_str(), proto.c_str(),
+            MDNS_LOOKUP_ACTIVE_TIME_MAX_MSEC, 1, &results);
 
-        if (results)
+    {
+        // Check if we need to (re-)enable advertising and disable the STA
+        // interface.
+        OSMutexLock locker(&lock_);
+        if (mdnsAdvInhibited_)
         {
-            // Only return one result.
-            struct addrinfo *addr_info =
-                (struct addrinfo*)memp_malloc(MEMP_NETDB);
-            if (addr_info == nullptr)
+            // We do not need to check for valid staIface_ or the
+            // mdnsAdvInhibitSta_ flag since we can only get here if it both
```

Comment:
Is `mdnsAdvInhibitSta_` the correct reference here? It seems that we have two vars referenced that have similar comments on usage/dependency.